File: tests/R/test_golden.R
# test_golden.R : Regression tests (golden files)

# Run the full pipeline with all the real data and verifies that
# results are identical to the reference files in tests/golden/.

# these tests are slow (few minutes), run:
#   testthat::test_file("tests/R/test_golden.R")
#
#To exclude when running all the tests:
#   testthat::test_dir("tests/R/", filter = "^(?!golden)")

source("helpers.R")

# ==== PIPELINE PARAMETRES ====

INHIBITION_THRESHOLD   <- 0.5
RATIO_THRESHOLD        <- 0.5
PROBABILITY_THRESHOLD  <- 0.75


# ==== HELPER: READ GOLDEN CSV ====

read_golden <- function(cell_line, sheet_name) {
  path <- file.path(GOLDEN_DIR, paste0(cell_line, "_", sheet_name, ".csv"))
  read.csv(path, header = FALSE, stringsAsFactors = FALSE)
}


# ==== HELPER: PREPARE AND EXECUTE A PIPELINE DIRECTORY ====

make_golden_run_dir <- function(cell_line) {
  run_dir <- tempfile(); dir.create(run_dir)
  file.copy(file.path(CODE_DIR, paste0(cell_line, ".xlsm")),
            file.path(run_dir, paste0(cell_line, ".xlsm")))
  req_dst <- file.path(run_dir, "requiredData"); dir.create(req_dst)
  file.copy(list.files(file.path(CODE_DIR, "requiredData"), full.names = TRUE), req_dst)
  run_dir
}

run_pipeline <- function(run_dir, cell_line) {
  old_wd <- getwd()
  tryCatch({
    setwd(run_dir)
    GetExpectancyOfBeingDownstreamTarget(
      inhibitionThreshold  = INHIBITION_THRESHOLD,
      ratioThreshold       = RATIO_THRESHOLD,
      probabilityThreshold = PROBABILITY_THRESHOLD,
      cellLinesFiles       = c(paste0(cell_line, ".xlsm"))
    )
  }, finally = setwd(old_wd))
}


# ==== RUNS PIPELINE WITH REAL DATA ====

run_dir_mcf7   <- make_golden_run_dir("MCF7");   run_pipeline(run_dir_mcf7,   "MCF7")
run_dir_hl60   <- make_golden_run_dir("HL60");   run_pipeline(run_dir_hl60,   "HL60")
run_dir_ntera2 <- make_golden_run_dir("NTERA2"); run_pipeline(run_dir_ntera2, "NTERA2")

# Load resulting workbooks
wb_mcf7   <- loadWorkbook(file.path(run_dir_mcf7,   "MCF7.xlsm"),   keepVBA = FALSE)
wb_hl60   <- loadWorkbook(file.path(run_dir_hl60,   "HL60.xlsm"),   keepVBA = FALSE)
wb_ntera2 <- loadWorkbook(file.path(run_dir_ntera2, "NTERA2.xlsm"), keepVBA = FALSE)


#==== TESTS: PutativeKinaseSubstrates ====

test_that("MCF7: number of Kinases with susbtrates that match golden", {
  golden <- read_golden("MCF7", "PutativeKinaseSubstrates")
  actual <- readWorkbook(wb_mcf7, sheet = "PutativeKinaseSubstrates", colNames = FALSE)

  golden_n <- sum(suppressWarnings(as.numeric(as.character(golden[-1, 2]))) > 0,
                  na.rm = TRUE)
  actual_n <- sum(suppressWarnings(as.numeric(as.character(actual[-1, 2]))) > 0,
                  na.rm = TRUE)

  expect_equal(actual_n, golden_n,
               label = paste("MCF7: kinases with substrates:", actual_n,
                             "vs", golden_n, "(golden)"))
})

test_that("HL60: number of kinases with substrates that match golden", {
  golden <- read_golden("HL60", "PutativeKinaseSubstrates")
  actual <- readWorkbook(wb_hl60, sheet = "PutativeKinaseSubstrates", colNames = FALSE)

  golden_n <- sum(suppressWarnings(as.numeric(as.character(golden[-1, 2]))) > 0,
                  na.rm = TRUE)
  actual_n <- sum(suppressWarnings(as.numeric(as.character(actual[-1, 2]))) > 0,
                  na.rm = TRUE)

  expect_equal(actual_n, golden_n)
})


# ==== TESTS: nodes.edges ====

test_that("MCF7: number of kinase pairs with edges that match golden", {
  golden <- read_golden("MCF7", "nodes.edges")
  actual <- readWorkbook(wb_mcf7, sheet = "nodes.edges", colNames = FALSE)

  golden_pares <- sum(!is.na(golden[-1, 1]) & golden[-1, 1] != "")
  actual_pares <- sum(!is.na(actual[-1, 1]) & actual[-1, 1] != "")

  expect_equal(actual_pares, golden_pares,
               label = paste("MCF7 pairs:", actual_pares, "vs", golden_pares, "(golden)"))
})

test_that("HL60: number of kinase pairs with edges that match golden", {
  golden <- read_golden("HL60", "nodes.edges")
  actual <- readWorkbook(wb_hl60, sheet = "nodes.edges", colNames = FALSE)

  golden_pares <- sum(!is.na(golden[-1, 1]) & golden[-1, 1] != "")
  actual_pares <- sum(!is.na(actual[-1, 1]) & actual[-1, 1] != "")

  expect_equal(actual_pares, golden_pares)
})


# ==== TESTS: CORRELATIONS (first row of MCF7) ====

test_that("MCF7: correlations first row that match golden (tolerance 1e-6)", {
  golden <- read_golden("MCF7", "corrPPwithKinases")
  actual <- readWorkbook(wb_mcf7, sheet = "corrPPwithKinases", colNames = FALSE)

  # Compare row 2 (first psite)) column to column
  golden_row <- suppressWarnings(as.numeric(as.character(golden[2, -1])))
  actual_row <- suppressWarnings(as.numeric(as.character(actual[2, -1])))

  expect_equal(length(actual_row), length(golden_row),
               label = "Column number different from corrPPwithKinases")

  for (j in seq_along(golden_row)) {
    if (!is.na(golden_row[j])) {
      expect_equal(actual_row[j], golden_row[j], tolerance = 1e-6,
                   label = paste("corrPPwithKinases col", j))
    }
  }
})


# ==== TESTS: TOP KINASE SUBSTRATES ====

test_that("MCF7: substrates of the kinase with more substrates that match golden", {
  golden <- read_golden("MCF7", "PutativeKinaseSubstrates")
  actual <- readWorkbook(wb_mcf7, sheet = "PutativeKinaseSubstrates", colNames = FALSE)

  # Find top kinase in golden
  golden_n    <- suppressWarnings(as.numeric(as.character(golden[-1, 2])))
  top_idx     <- which.max(golden_n)
  top_kinase  <- as.character(golden[top_idx + 1, 1])
  top_n       <- golden_n[top_idx]

  #Search in current
  actual_fila <- actual[as.character(actual[, 1]) == top_kinase, ]
  actual_n    <- suppressWarnings(as.numeric(as.character(actual_fila[1, 2])))

  expect_equal(actual_n, top_n,
               label = paste("Substrates of", top_kinase, ":", actual_n,
                             "vs", top_n, "(golden)"))
})


# ==== TESTS: NTERA2 ====

test_that("NTERA2: The number of kinases with substrates matches golden", {
  golden <- read_golden("NTERA2", "PutativeKinaseSubstrates")
  actual <- readWorkbook(wb_ntera2, sheet = "PutativeKinaseSubstrates", colNames = FALSE)

  golden_n <- sum(suppressWarnings(as.numeric(as.character(golden[-1, 2]))) > 0,
                  na.rm = TRUE)
  actual_n <- sum(suppressWarnings(as.numeric(as.character(actual[-1, 2]))) > 0,
                  na.rm = TRUE)

  expect_equal(actual_n, golden_n,
               label = paste("NTERA2: kinases with substrates:", actual_n,
                             "vs", golden_n, "(golden)"))
})

test_that("NTERA2: number of kinase pairs with edges that match golden", {
  golden <- read_golden("NTERA2", "nodes.edges")
  actual <- readWorkbook(wb_ntera2, sheet = "nodes.edges", colNames = FALSE)

  golden_pares <- sum(!is.na(golden[-1, 1]) & golden[-1, 1] != "")
  actual_pares <- sum(!is.na(actual[-1, 1]) & actual[-1, 1] != "")

  expect_equal(actual_pares, golden_pares,
               label = paste("NTERA2 pairs:", actual_pares, "vs", golden_pares, "(golden)"))
})

test_that("NTERA2: first row correlations that match golden (tolerancia 1e-6)", {
  golden <- read_golden("NTERA2", "corrPPwithKinases")
  actual <- readWorkbook(wb_ntera2, sheet = "corrPPwithKinases", colNames = FALSE)

  golden_row <- suppressWarnings(as.numeric(as.character(golden[2, -1])))
  actual_row <- suppressWarnings(as.numeric(as.character(actual[2, -1])))

  expect_equal(length(actual_row), length(golden_row),
               label = "different column number in NTERA2 corrPPwithKinases")

  for (j in seq_along(golden_row)) {
    if (!is.na(golden_row[j])) {
      expect_equal(actual_row[j], golden_row[j], tolerance = 1e-6,
                   label = paste("NTERA2 corrPPwithKinases col", j))
    }
  }
})

test_that("NTERA2: substrates of the kinase with more substrates that match golden", {
  golden <- read_golden("NTERA2", "PutativeKinaseSubstrates")
  actual <- readWorkbook(wb_ntera2, sheet = "PutativeKinaseSubstrates", colNames = FALSE)

  golden_n    <- suppressWarnings(as.numeric(as.character(golden[-1, 2])))
  top_idx     <- which.max(golden_n)
  top_kinase  <- as.character(golden[top_idx + 1, 1])
  top_n       <- golden_n[top_idx]

  actual_fila <- actual[as.character(actual[, 1]) == top_kinase, ]
  actual_n    <- suppressWarnings(as.numeric(as.character(actual_fila[1, 2])))

  expect_equal(actual_n, top_n,
               label = paste("NTERA2 substrates of", top_kinase, ":", actual_n,
                             "vs", top_n, "(golden)"))
})

File: tests/python/test_golden.py
"""
test_golden.py - regression tests (golden files)

run the full pipeline with the real data (MCF7.xlsm, HL60.xlsm and NTERA2.xlsm)
and verify results are identical to the stored reference results in tests/golden/

These tests are useful to:
  - They guarantee that the code will produce the same results on any machine or future version
  - If someone modifies the algorithm, these test will fail an alert you

*Note:These tests are slow bacuse they process bigger datsets
Run:
    pytest tests/python/test_golden.py -v

Or exclude in quick runs:
    pytest tests/python/ -v -m "not slow"
"""

import os
import csv
import math
import shutil
import pytest
import openpyxl

from ebdtFunctions import GetExpectancyOfBeingDownstreamTarget


# ==== PARAMETERS OF THE PIPELINE (must match with those used when generating golden) ====

INHIBITION_THRESHOLD   = 0.5
RATIO_THRESHOLD        = 0.5
PROBABILITY_THRESHOLD  = 0.75


# ==== FIXTURE: EXECUTED REAL PIPELINE ====

@pytest.fixture(scope="module")
def golden_results(tmp_path_factory):
    """
    Executes real pipeline with real MCF7, HL60 and NTERA2.
    returns dict: {'MCF7': workbook, 'HL60': workbook, 'NTERA2': workbook}
    """
    from conftest import CODE_DIR, GOLDEN_DIR
    run_dir = str(tmp_path_factory.mktemp("golden_run"))

    for cell_line in ["MCF7", "HL60", "NTERA2"]:
        shutil.copy(
            os.path.join(CODE_DIR, f"{cell_line}.xlsm"),
            os.path.join(run_dir, f"{cell_line}.xlsm")
        )
    shutil.copytree(
        os.path.join(CODE_DIR, "requiredData"),
        os.path.join(run_dir, "requiredData")
    )

    old_cwd = os.getcwd()
    try:
        os.chdir(run_dir)
        GetExpectancyOfBeingDownstreamTarget(
            inhibitionThreshold  = INHIBITION_THRESHOLD,
            ratioThreshold       = RATIO_THRESHOLD,
            probabilityThreshold = PROBABILITY_THRESHOLD,
            cellLinesFiles       = ["MCF7.xlsm", "HL60.xlsm", "NTERA2.xlsm"],
        )
    finally:
        os.chdir(old_cwd)

    return {
        cell_line: openpyxl.load_workbook(
            os.path.join(run_dir, f"{cell_line}.xlsm"),
            read_only=True,
            keep_vba=False
        )
        for cell_line in ["MCF7", "HL60", "NTERA2"]
    }


# ==== HELPER: READ GOLDEN CSV ====

def _read_golden(filename):
    from conftest import GOLDEN_DIR
    golden_dir = GOLDEN_DIR
    #reads a golden file CSV and returns a list of lists
    path = os.path.join(golden_dir, filename)
    with open(path, encoding="utf-8") as f:
        return list(csv.reader(f))


def _sheet_to_rows(workbook, sheet_name):
    #Extracts all the rows of a sheet as a list of lists
    ws = workbook[sheet_name]
    return [list(row) for row in ws.iter_rows(values_only=True)]


# ==== TESTS: PutativeKinaseSubstrates ====

@pytest.mark.slow
@pytest.mark.parametrize("cell_line", ["MCF7", "HL60", "NTERA2"])
def test_golden_numero_quinasas_con_sustratos(cell_line, golden_results):
    #the number of kinases with at least one putative substrate must be identical to the one in the golden file
    
    golden = _read_golden(f"{cell_line}_PutativeKinaseSubstrates.csv")
    actual_wb = golden_results[cell_line]
    actual = _sheet_to_rows(actual_wb, "PutativeKinaseSubstrates")

    # Counts kinases with n>0
    golden_con_sust = sum(1 for r in golden[1:] if r[1] and int(float(r[1])) > 0)
    actual_con_sust = sum(1 for r in actual[1:] if r[1] is not None and int(float(r[1])) > 0)

    assert actual_con_sust == golden_con_sust, (
        f"{cell_line}: {actual_con_sust} kinases with substrates "
        f"(e {golden_con_sust})"
    )


@pytest.mark.slow
@pytest.mark.parametrize("cell_line", ["MCF7", "HL60", "NTERA2"])
def test_golden_sustratos_de_quinasa_top(cell_line, golden_results):
    """
    Substrates of the kinase with more substrates must be identical
    to the ones in the golden file (exact reproducibility test).
    """
    golden = _read_golden(f"{cell_line}_PutativeKinaseSubstrates.csv")
    actual_wb = golden_results[cell_line]
    actual = _sheet_to_rows(actual_wb, "PutativeKinaseSubstrates")

    # finds the kinase with more substrates in the golden
    golden_data = [(r[0], int(float(r[1])), r[2]) for r in golden[1:] if r[1]]
    golden_data.sort(key=lambda x: -x[1])
    top_kinase, top_n, top_subs_golden = golden_data[0]

    # search this kinase in the current result
    actual_row = next((r for r in actual[1:] if r[0] == top_kinase), None)
    assert actual_row is not None, f"Not found {top_kinase} in result"

    assert int(float(actual_row[1])) == top_n, (
        f"{cell_line} {top_kinase}: {actual_row[1]} substrates ({top_n})"
    )


# ==== TESTS: nodes.edges ====

@pytest.mark.slow
@pytest.mark.parametrize("cell_line", ["MCF7", "HL60", "NTERA2"])
def test_golden_numero_pares_kinasas(cell_line, golden_results):
    """
    the number of kinase pairs with more shared substrates must be
    identical to the one in the golden file.
    """
    golden = _read_golden(f"{cell_line}_nodes.edges.csv")
    actual_wb = golden_results[cell_line]
    actual = _sheet_to_rows(actual_wb, "nodes.edges")

    # -1 for the header 
    golden_pares = len([r for r in golden[1:] if r[0]])
    actual_pares = len([r for r in actual[1:] if r[0] is not None])

    assert actual_pares == golden_pares, (
        f"{cell_line}: {actual_pares} pairs ({golden_pares})"
    )


# ==== TESTS: corrPPwithKinases ====

@pytest.mark.slow
@pytest.mark.parametrize("cell_line", ["MCF7", "HL60", "NTERA2"])
def test_golden_correlaciones_primera_fila(cell_line, golden_results):
    """
    Correlations of the first row of psites must be
    numerically equal to the ones in the golden file (tolerance 1e-6).
    """
    golden = _read_golden(f"{cell_line}_corrPPwithKinases.csv")
    actual_wb = golden_results[cell_line]
    actual = _sheet_to_rows(actual_wb, "corrPPwithKinases")

    golden_row = golden[1]
    actual_row = actual[1]

    assert len(golden_row) == len(actual_row), (
        f"{cell_line}: number of different columns: {len(actual_row)} vs {len(golden_row)}"
    )

    for col_idx, (g_val, a_val) in enumerate(zip(golden_row[1:], actual_row[1:]), start=1):
        if g_val == "" or g_val is None:
            assert a_val is None or a_val == ""
        else:
            g_float = float(g_val) if g_val else 0.0
            a_float = float(a_val) if a_val is not None else 0.0
            assert math.isclose(g_float, a_float, rel_tol=1e-6, abs_tol=1e-9), (
                f"{cell_line} col {col_idx}: {a_float} != {g_float} (golden)"
            )


@pytest.mark.slow
def test_golden_MCF7_correlaciones_primera_fila(golden_results):
    """
    Correlations of the first row of psites (MCF7) must be
    numerically equal to the ones in the golden file (tolerance 1e-6).
    """
    golden = _read_golden("MCF7_corrPPwithKinases.csv")
    actual_wb = golden_results["MCF7"]
    actual = _sheet_to_rows(actual_wb, "corrPPwithKinases")

    # Comparare row 2 (first psite, index 1 in the list)
    golden_row = golden[1]
    actual_row = actual[1]

    assert len(golden_row) == len(actual_row), (
        f"Number of different columns: {len(actual_row)} vs {len(golden_row)}"
    )

    for col_idx, (g_val, a_val) in enumerate(zip(golden_row[1:], actual_row[1:]), start=1):
        if g_val == "" or g_val is None:
            assert a_val is None or a_val == ""
        else:
            g_float = float(g_val) if g_val else 0.0
            a_float = float(a_val) if a_val is not None else 0.0
            assert math.isclose(g_float, a_float, rel_tol=1e-6, abs_tol=1e-9), (
                f"Col {col_idx}: {a_float} != {g_float} (golden)"
            )
